pkg/services/service.go

```diff
@@ -0,0 +1,228 @@
+//
+//	defer Unhealthy(fmt.Errorf("foo bar: %i", err))()
+//
+// See SetUnhealthy for an alternative API.
+func (w *Engine) Unhealthy(err error) func() {
+	cond := uuid.NewString()
+	w.SetUnhealthy(cond, err)
+	return func() { w.SetHealthy(cond) }
+}
+
+func (w *Engine) healthy() error {
+	w.condsMu.RLock()
+	errs := maps.Values(w.conds)
+	w.condsMu.RUnlock()
+	return errors.Join(errs...)
+}
+
+// Config is a service configuration.
+type Config struct {
+	// Name is required. It will be logged shorthand on Start and Close, and appended to the logger name.
+	// It must be unique among services sharing the same logger, in order to ensure uniqueness of the fully qualified name.
+	Name string
+	// SubServices is an optional set of dependent Services to Start and Close along with this one.
+	NewSubServices func(logger.Logger) []Service
+	// Start is an optional hook called after starting SubServices.
+	Start func(context.Context) error
+	// Close is an optional hook called before closing SubServices.
+	Close func() error //TODO examples
+}
+
+type service struct {
+	StateMachine
+	cfg  Config
+	eng  Engine
+	subs []Service
+}
+
+// NewService returns a new Service defined by Config, and an Engine for managing health, goroutines, and logging.
+//   - You *should* embed the Service, in order to inherit the methods.
+//   - You *should not* embed the Engine. Use an unexported field instead.
+func (c Config) NewService(lggr logger.Logger) (Service, *Engine) {
+	s := c.new(lggr)
+	return s, &s.eng
+}
+
+// NewService is equivalent to calling Config.NewService and discarding the Engine.
+// This can be inlined at instantiation time for cases w/o self-references (no Config.Start or Config.SubServices).
+// TODO is this really useful?
+func NewService(lggr logger.Logger, name string) Service {
+	return Config{Name: name}.new(lggr)
+}
+
+func (c Config) new(lggr logger.Logger) *service {
+	lggr = logger.Sugared(logger.Named(lggr, c.Name))
+	s := &service{
+		cfg: c,
+		eng: Engine{
+			StopChan:      make(StopChan),
+			SugaredLogger: logger.Sugared(lggr),
+			conds:         make(map[string]error),
+		},
+	}
+	//TODO s.eng.somethingFn = s.StateMachine.ErrorBuffer.Append
+	if c.NewSubServices != nil {
+		s.subs = c.NewSubServices(lggr)
+	}
+	return s
+}
+
+// Ready implements [HealthReporter.Ready] and overrides and extends [utils.StartStopOnce.Ready()] to include [Config.SubServices]
+// readiness as well.
+func (s *service) Ready() (err error) {
+	err = s.StateMachine.Ready()
+	for _, sub := range s.subs {
+		err = errors.Join(err, sub.Ready())
+	}
+	return
+}
+
+// Healthy overrides [StateMachine.Healthy] and extends it to include Engine errors as well.
+// Do not override this method in your service. Instead, report errors via the Engine.
+func (s *service) Healthy() (err error) {
+	err = s.StateMachine.Healthy()
+	if err == nil {
+		err = s.eng.healthy()
+	}
+	return
+}
+
+func (s *service) HealthReport() map[string]error {
```

Comment:
As a fail-safe, we could track when this is called, and if it is never called, or not called often enough, then we could periodically emit to logs instead :thinking:

Reply:
I'm still open to this, but as a follow-up after this PR.